File: T34-Tuplespaces-2025/Client-Python/client_service.py
import sys
import grpc
sys.path.insert(1, '../Contract/target/generated-sources/protobuf/python')

from TupleSpaces_pb2 import PutRequest, ReadRequest, TakeRequest, getTupleSpacesStateRequest
from TupleSpaces_pb2_grpc import TupleSpacesStub
from client_debug import ClientDebug

class ClientService:
    # Atributos da classe
    def __init__(self, host_port: str, client_id: int):
        self.client_id = client_id
        self.channel = grpc.insecure_channel(host_port)
        self.stub = TupleSpacesStub(self.channel)

    def put(self, tuple: str, delays: list[int]):
        ClientDebug.debug(f"Client {self.client_id}: put({tuple}) delays={delays}")
        try:
            request = PutRequest(newTuple=tuple)
            metadata = [("client-id", str(self.client_id))] + [(f"delay{i+1}", str(delay)) for i, delay in enumerate(delays)]
            response = self.stub.put(request, metadata=metadata)
            return True
        except grpc.RpcError as e:
            status_code = e.code()
            ClientDebug.debug(f"Error: CODE {status_code}")
            return False

    def read(self, pattern: str, delays: list[int]):
        ClientDebug.debug(f"Client {self.client_id}: read({pattern}) delays={delays}")
        try:
            request = ReadRequest(searchPattern=pattern)
            metadata = [("client-id", str(self.client_id))] + [(f"delay{i+1}", str(delay)) for i, delay in enumerate(delays)]
            response = self.stub.read(request, metadata=metadata)
            return response.result
        except grpc.RpcError as e:
            status_code = e.code()
            ClientDebug.debug(f"Error: CODE {status_code}")
            return ""

    def take(self, pattern: str, delays: list[int]):
        ClientDebug.debug(f"Client {self.client_id}: take({pattern}) delays={delays}")
        try:
            request = TakeRequest(searchPattern=pattern)
            metadata = [("client-id", str(self.client_id))] + [(f"delay{i+1}", str(delay)) for i, delay in enumerate(delays)]
            response = self.stub.take(request, metadata=metadata)
            return response.result
        except grpc.RpcError as e:
            status_code = e.code()
            ClientDebug.debug(f"Error: CODE {status_code}")
            return ""

    def get_tuple_spaces_state(self, delays: list[int]):
        ClientDebug.debug(f"Client {self.client_id}: getTupleSpacesState() delays={delays}")
        try:
            request = getTupleSpacesStateRequest()
            metadata = [("client-id", str(self.client_id))] +  [(f"delay{i+1}", str(delay)) for i, delay in enumerate(delays)]
            response = self.stub.getTupleSpacesState(request, metadata=metadata)
            return response.tuple
        except grpc.RpcError as e:
            status_code = e.code()
            ClientDebug.debug(f"Error: CODE {status_code}")
            return None


    def shutdown(self):
        self.channel.close()

File: T34-Tuplespaces-2025/Client-Java/src/main/java/pt/ulisboa/tecnico/tuplespaces/client/grpc/ClientService.java
package pt.ulisboa.tecnico.tuplespaces.client.grpc;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Metadata;
import io.grpc.StatusRuntimeException;
import io.grpc.stub.MetadataUtils;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TupleSpacesGrpc;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TupleSpacesGrpc.TupleSpacesBlockingStub;
import pt.ulisboa.tecnico.tuplespaces.centralized.contract.TupleSpacesOuterClass.*;
import static pt.ulisboa.tecnico.tuplespaces.client.ClientDebug.debug;

import java.util.List;

public class ClientService {
    private final int client_id;
    private final ManagedChannel channel;

    public ClientService(String host_port, int client_id) {
        this.client_id = client_id;
        this.channel = ManagedChannelBuilder.forTarget(host_port)
                .usePlaintext()
                .build();
    }

    /*
     * Sends a put() request to the front-end to insert the specified tuple.
     * This operation adds the tuple to the replicated tuple space.
     * 
     * If a delays array is provided, it will apply artificial delays (in seconds)
     * to each replica before processing the request.
     * 
     * @param tuple The tuple to be inserted, in the format "<field1,field2,...>".
     * @param delays An array of 3 integers representing delays for replicas 0, 1, and 2.
     * @return true if the operation was successful, false if an error occurred at the front-end or any replica.
     */
    public boolean put(String tuple, Integer[] delays) {
        debug(String.format("Client %d: put(%s) delays=%s", client_id, tuple, java.util.Arrays.toString(delays)));
    
        try {
            getBlockingStub(delays).put(PutRequest.newBuilder().setNewTuple(tuple).build());
            return true;
        } catch (StatusRuntimeException e) {
            System.err.println(e.getStatus().getDescription() + System.lineSeparator());
            return false;
        }
    }

    // Overload with default delays of 0s.
    public boolean put(String tuple) {
        return put(tuple, new Integer[]{0, 0, 0});
    }

    /**
     * Sends a read() request to the front-end to search for a tuple matching the provided pattern.
     * This operation will return one matching tuple, if available, without removing it from the tuple space.
     * The client will block until a matching tuple is found.
     * 
     * Optional delays (in seconds) can be applied to replicas during processing.
     * 
     * @param pattern The search pattern (can be a regular expression) used to match tuples.
     * @param delays An array of 3 integers representing delays for replicas 0, 1, and 2.
     * @return The first matching tuple found, or null if an error occurred.
     */
    public String read(String pattern, Integer[] delays) {
        debug(String.format("Client %d: read(%s) delays=%s", client_id, pattern, java.util.Arrays.toString(delays)));
        try {
            ReadResponse response = getBlockingStub(delays).read(ReadRequest.newBuilder().setSearchPattern(pattern).build());
            return response.getResult();
        } catch (StatusRuntimeException e) {
            System.err.println(e.getStatus().getDescription() + System.lineSeparator());
            return null;
        }
    }

    public String read(String pattern) {
        return read(pattern, new Integer[]{0, 0, 0});
    }


    /**
     * Sends a take() request to the front-end to search and remove a tuple matching the provided pattern. 
     * Optional delays (in seconds) can be applied to replicas during processing.
     * 
     * @param pattern The search pattern 
     * @param delays An array of 3 integers representing delays for replicas 0, 1, and 2.
     * @return The tuple that was removed, or null if an error occurred.
     */
    public String take(String pattern, Integer[] delays) {
        debug(String.format("Client %d: take(%s) delays=%s", client_id, pattern, java.util.Arrays.toString(delays)));
        try {
            TakeResponse response = getBlockingStub(delays).take(TakeRequest.newBuilder().setSearchPattern(pattern).build());
            return response.getResult();
        } catch (StatusRuntimeException e) {
            System.err.println(e.getStatus().getDescription() + System.lineSeparator());
            return null;
        }
    }

    public String take(String pattern) {
        return take(pattern, new Integer[]{0, 0, 0});
    }


    /**
     * Sends a getTupleSpacesState() request to retrieve the current state (all tuples)
     * from the replicated tuple space servers.
     * 
     * Optional delays (in seconds) can be applied to replicas during processing.
     * 
     * @param delays An array of 3 integers representing delays for replicas 0, 1, and 2.
     * @return A list of all tuples from the tuple space, or null if an error occurred.
     */
    public List<String> getTupleSpacesState(Integer[] delays) {
        debug(String.format("Client %d: getTupleSpacesState() delays=%s", client_id, java.util.Arrays.toString(delays)));
        try {
            getTupleSpacesStateRequest request = getTupleSpacesStateRequest.newBuilder().build();

            getTupleSpacesStateResponse response = getBlockingStub(delays)
                .getTupleSpacesState(request);

            return response.getTupleList();
        } catch (StatusRuntimeException e) {
            System.err.println(e.getStatus().getDescription() + System.lineSeparator());
            return null;
        }
    }

    public List<String> getTupleSpacesState() {
        return getTupleSpacesState(new Integer[]{0, 0, 0});
    }

    
    
    public void shutdown() {
        channel.shutdown();
    }

    /**
     * Creates a new blocking stub for communicating with the front-end,
     * attaching gRPC metadata containing the client ID and delays for each replica.
     * 
     * The delays array should contain 3 elements, each representing the artificial delay (in seconds)
     * to be applied by replica 0, 1, and 2, respectively.
     * 
     * @param delays An array of 3 integers for delays (replica0, replica1, replica2).
     * @return A blocking stub with metadata attached.
     */
    private TupleSpacesBlockingStub getBlockingStub(Integer[] delays) {
        Metadata metadata = new Metadata();
        Metadata.Key<String> clientIdKey = Metadata.Key.of("client-id", Metadata.ASCII_STRING_MARSHALLER);
        metadata.put(clientIdKey, String.valueOf(client_id));
    
        for (int i = 0; i < delays.length; i++) {
            Metadata.Key<String> delayKey = Metadata.Key.of("delay" + (i + 1), Metadata.ASCII_STRING_MARSHALLER);
            metadata.put(delayKey, String.valueOf(delays[i]));
        }
    
        return TupleSpacesGrpc.newBlockingStub(channel).withInterceptors(MetadataUtils.newAttachHeadersInterceptor(metadata));
    }
    
}